servers/zms/src/main/java/com/yahoo/athenz/zms/notification/NotificationManager.java

```diff
@@ -0,0 +1,162 @@
+        }
+    }
+
+    public NotificationManager(DBService dbService, NotificationServiceFactory notificationServiceFactory) {
+
+        this.dbService = dbService;
+        this.notificationServiceFactory = notificationServiceFactory;
+        notificationService = this.notificationServiceFactory.create();
+
+        init();
+    }
+
+    void shutdown() {
+        if (scheduledExecutor != null) {
+            scheduledExecutor.shutdownNow();
+        }
+    }
+
+    public void generateAndSendPostPutMembershipNotification(String domain, String org, Boolean auditEnabled, Boolean selfserve, Map<String, String> details) {
+
+        Set<String> recipients = new HashSet<>();
+        if (auditEnabled == Boolean.TRUE) {
+            //get recipient role(s) from audit domain
+            Role domainRole = dbService.getRole(ZMSConsts.SYS_AUTH_AUDIT_DOMAIN, ZMSConsts.AUDIT_APPROVER_ROLE_PREFIX + org + "." + domain, false, true, false);
+            Role orgRole = dbService.getRole(ZMSConsts.SYS_AUTH_AUDIT_DOMAIN, ZMSConsts.AUDIT_APPROVER_ROLE_PREFIX + org, false, true, false);
+            if (domainRole != null) {
+                recipients.addAll(domainRole.getRoleMembers().stream()
+                        .map(RoleMember::getMemberName).collect(Collectors.toSet()));
+            }
+            if (orgRole != null) {
+                recipients.addAll(orgRole.getRoleMembers().stream()
+                        .map(RoleMember::getMemberName).collect(Collectors.toSet()));
+            }
+        } else if (selfserve == Boolean.TRUE) {
+            // get admin role from the request domain
+            Role adminRole = dbService.getRole(domain, "admin", false, true, false);
+            recipients.addAll(adminRole.getRoleMembers().stream()
+                    .map(RoleMember::getMemberName).collect(Collectors.toSet()));
+        }
+        Notification notification = createNotification(ZMSConsts.NOTIFICATION_TYPE_MEMBERSHIP_APPROVAL, recipients, details);
+        notificationService.notify(notification);
+    }
+
+    public Notification createNotification(String notificationType, Set<String> recipients, Map<String, String> details) {
+        String recDomain, recRole;
+        AthenzDomain domain;
+        Notification notification = new Notification(notificationType);
+        notification.setDetails(details);
+        for (String recipient : recipients) {
+            if (recipient.contains(":role.")) {
+                //recipient is of type role. Extract role members
+                recDomain = recipient.substring(0, recipient.indexOf(":"));
+                recRole = recipient.substring(recipient.indexOf(":") + 6);
+                domain = dbService.getAthenzDomain(recDomain, false);
+                for (Role role : domain.getRoles()) {
+                    if (role.getName().equalsIgnoreCase(domain.getName() + ":role." + recRole)) {
```

Comment:
isn't domain.getName() + ":role." + recRole is the same as recipient?

and if that's the case why are we extracting the recRole value?

any reason for equalsIgnoreCase since internally we already lower cased everything?